### PR TITLE
Add support for Infineon XMC4500 SoC

### PR DESCRIPTION
**Pull request by parthitce**
This PR adds support for Infineon XMC4500 with development board relax kit. Currently the west module dependency points to https://github.com/linumiz/hal_infineon.

hal_infineon RFC request here: https://github.com/zephyrproject-rtos/zephyr/issues/24554

### Support:
- Initial revison adds support only for UART in polling mode
- Uses DSRAM1 as default SRAM (64KB)
- Clock defaults to 120MHz

### WIP
- Clock Control, GPIO and PINMUX support work is in progress and will be submitted in separate PR once the basic structure for infineon is accepted
- RAM: 160KB is split into PSRAM1 64KB, DSRAM1 64KB and DSRAM2 32KB (communication RAM used for USB and Ethernet). CONFIG_BT_STM32_IPM in linker.ld will be extended to address this RAM split with the same name (SRAM1 & SRAM2).

### CODE REVIEW

File: soc/arm/infineon_xmc/4xxx/soc.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#include <soc.h>
+
+#define PMU_FLASH_WS		(0x3U)
+
+void z_platform_init(void)
+{
+	u32_t temp;
+
+	/* Enable FPU */
+	SCB->CPACR |= ((3UL << 10*2) | (3UL << 11*2));
```
Comment (galak) [inline soc/arm/infineon_xmc/4xxx/soc.c:20]: Can we use CPACR defines from cmsis.h so this isn't all magic numbers.  Should this also be wrapped in a CONFIG_FLOAT

**Review by galak — COMMENTED** [review]
At first glance this looks good.

Are you able to commit to maintaining this SoC port, drivers, etc associated with it?

**Comment by parthitce** [comment]
> At first glance this looks good.
> 
> Are you able to commit to maintaining this SoC port, drivers, etc associated with it?

Yes, I will be able to maintain this port. Thanks.

File: soc/arm/infineon_xmc/4xxx/soc.c (hunk 1 above)
Comment (parthitce) [reply soc/arm/infineon_xmc/4xxx/soc.c]: Sure, adapted it. Thanks.

File: soc/arm/infineon_xmc/4xxx/soc.c [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#define PMU_FLASH_WS		(0x3U)
+
+void z_platform_init(void)
+{
+	u32_t temp;
+
+#if defined(CONFIG_FLOAT)
+	/* Enable FPU */
+	SCB->CPACR |= CPACR_CP10_Msk | CPACR_CP11_Msk;
+#endif
```
Comment (stephanosio) [inline soc/arm/infineon_xmc/4xxx/soc.c:23]: Do we have to allow FPU coproc access here (i.e. before `prep_c`)? If not, this should be removed since the ARM arch port does this in `prep_c`

https://github.com/zephyrproject-rtos/zephyr/blob/b39423e0e27cc08915b6f9dca1b4b57cff5e7c9e/arch/arm/core/aarch32/prep_c.c#L80-L87

Comment (parthitce) [reply soc/arm/infineon_xmc/4xxx/soc.c]: @stephanosio Thanks for pointing this. It doesn't needs to be earlier than prep_c, removed FPU handling in soc.

**Comment by carlescufi** [comment]
@parthitce we should be able to create the module repo soon. In the meantime, can you make sure this is rebased and up to date with the latest DT changes?

File: soc/arm/infineon_xmc/4xxx/Kconfig.defconfig.series [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# XMC 4xxx series configuration options
+# Copyright (c) 2020 Linumiz
+# Author: Parthiban Nallathambi <parthiban@linumiz.com>
+
+if SOC_SERIES_XMC_4XXX
+
+config SOC_SERIES
+	default "4xxx"
+
+config FLOAT
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply soc/arm/infineon_xmc/4xxx/Kconfig.defconfig.series:12]: Yes, and by the way, this option should not be defined here as =y; it should either be defined in the Board Kconfig definition or in the module (e.g. driver) that requires that.

[the remark this answers is not in the document]
Comment (parthitce) [reply soc/arm/infineon_xmc/4xxx/Kconfig.defconfig.series:12]: @MaureenHelm @ioannisg Thanks for pointing it. If I understand correctly, CPU_HAS_FPU defines the presence of FPU in the SoC and it is selected in Kconfig.soc. Using FPU depends on the application itself and it's discouraged to define it Y in soc Kconfig?

[the remark this answers is not in the document]
Comment (ioannisg) [reply soc/arm/infineon_xmc/4xxx/Kconfig.defconfig.series:12]: Yes, correct, CONFIG_FPU and CONFIG_FPU_SHARING are application options, not related to the SoC configuration.

Note that they have just been renamed :)

File: boards/arm/relax_kit/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,82 @@
+.. _relax_kit:
+
+INFINEON RELAX-KIT
+##################
+
+Overview
+********
+
+TheXMC4500 Relax Kit designed to evaluate the capabilities of the XMC4500
```
Comment (ioannisg) [inline boards/arm/relax_kit/doc/index.rst:9]: is designed

File: boards/arm/relax_kit/doc/index.rst [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+Microcontroller. It contains High performance ARM Cortex-M4F which can run
```
Comment (ioannisg) [inline boards/arm/relax_kit/doc/index.rst:10]: nit: instead of "contains", "includes" or "is based on"

File: boards/arm/relax_kit/relax_kit_defconfig [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (ioannisg) [inline boards/arm/relax_kit/relax_kit_defconfig:1]: Since the CPU has ARM MPU, can you enable by default the ARM_MPU here? See how it is done in other defconfig of ARM boards.

File: soc/arm/infineon_xmc/4xxx/noinit.ld [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2020 Linumiz
+ * Author: Parthiban Nallathambi <parthiban@linumiz.com>
+ *
+ * Extracted from: infineon/XMCLib/devices/XMC4500/Source/GCC/XMC4500x1024.ld
+ */
+
+/* .no_init section contains chipid, SystemCoreClock and trimming data */
+KEEP(*(.no_init))
```
Comment (ioannisg) [inline soc/arm/infineon_xmc/4xxx/noinit.ld:11]: is this really needed?

File: soc/arm/infineon_xmc/4xxx/soc.c [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+#include <soc.h>
+
+#define PMU_FLASH_WS		(0x3U)
+
+void z_platform_init(void)
+{
+	u32_t temp;
+
+	/* Enable unaligned memory access - SCB_CCR.UNALIGN_TRP = 0 */
+	SCB->CCR &= ~(SCB_CCR_UNALIGN_TRP_Msk);
```
Comment (ioannisg) [inline soc/arm/infineon_xmc/4xxx/soc.c:20]: Why is this needed? Is it required temporary? AFAIK (maybe I am wrong though) this is not done for other cortex-m platforms. Must justify.

**Review by ioannisg — CHANGES_REQUESTED** [review]
Overall looking good
I added some comments, pls address

File: soc/arm/infineon_xmc/4xxx/Kconfig.defconfig.series (hunk 3 above)
[the remark this answers is not in the document]
Comment (parthitce) [reply soc/arm/infineon_xmc/4xxx/Kconfig.defconfig.series]: Thanks, removed.

File: boards/arm/relax_kit/doc/index.rst (hunk 4 above)
Comment (parthitce) [reply boards/arm/relax_kit/doc/index.rst]: Fixed, thanks

File: boards/arm/relax_kit/doc/index.rst (hunk 5 above)
Comment (parthitce) [reply boards/arm/relax_kit/doc/index.rst]: Fixed, thanks.

File: boards/arm/relax_kit/relax_kit_defconfig (hunk 6 above)
Comment (parthitce) [reply boards/arm/relax_kit/relax_kit_defconfig]: I have a compilation failure when enabled. Sure I will check this, thanks.

File: soc/arm/infineon_xmc/4xxx/noinit.ld (hunk 7 above)
Comment (parthitce) [reply soc/arm/infineon_xmc/4xxx/noinit.ld:11]: Yes, SoC clock initialization uses device specific routines defined [here](https://github.com/linumiz/hal_infineon/blob/master/XMCLib/devices/XMC4500/Source/system_XMC4500.c). SystemCoreClock variable is defined under this special section as [here](https://github.com/linumiz/hal_infineon/blob/master/XMCLib/devices/XMC4500/Source/system_XMC4500.c#L299).

Clock control is currently WIP [here](https://github.com/parthitce/zephyr/commit/fcbb3ca2a34dde97bd04cf7fb16fc8d1f3fa2e39#diff-1f590f20845a91f285809fb66b13b9b4), which will removed dependency with device specific init and probably SystemCoreClock moves into zephyr bss section and calculates the clock value based on Kconfig's.

File: soc/arm/infineon_xmc/4xxx/soc.c (hunk 8 above)
Comment (parthitce) [reply soc/arm/infineon_xmc/4xxx/soc.c]: Agree, it was temporary and shouldn't be here. Removed, thanks.

**Comment by parthitce** [comment]
Apart from the comments fix, I did modify 2 major things now in the latest push.

1. Board named is renamed from relax_kit to xmc45_relax_kit. This is because infineon releases relax_kit for all the XMC series.

2. Renamed the uart node name from uart to usic. To brief, serial communication interfaces in XMC SoC is handled by (Universal Serial Interface Channel (USIC). XMC4500 got 3 USIC instances with 2 channels each. USIC can be configured to use as UART, SPI, I2C, I2S. Now the parent dtsi defines only usic with channels and consumed as uart in the board specific dts file.

Thanks in advance for your re-review @ioannisg @MaureenHelm @galak @carlescufi @stephanosio

**Comment by carlescufi** [comment]
DNM until this points at our `hal_infineon` repo

**Comment by parthitce** [comment]
> DNM until this points at our `hal_infineon` repo

@carlescufi updated pointer in west.yml to https://github.com/zephyrproject-rtos/hal_infineon

File: boards/arm/relax_kit/relax_kit_defconfig (hunk 6 above)
Comment (ioannisg) [reply boards/arm/relax_kit/relax_kit_defconfig]: @parthitce  i think you need to include devicetree.h in your soc.h.

```
/* Add include for DTS generated information */
#include <devicetree.h>

```
 See, for example the Nordic SOCS.

Let me know if this solves your error

Comment (parthitce) [reply boards/arm/relax_kit/relax_kit_defconfig]: @ioannisg Yes, that solves the problem. Thank you.

File: soc/arm/infineon_xmc/4xxx/Kconfig.series [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+# Author: Parthiban Nallathambi <parthiban@linumiz.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_XMC_4XXX
+	bool "XMC 4xxx Series MCU"
+	select ARM
+	select CPU_CORTEX_M4
+	select CPU_CORTEX_M_HAS_DWT
+	select SOC_FAMILY_XMC
+	select CPU_HAS_ARM_MPU
```
Comment (ioannisg) [inline soc/arm/infineon_xmc/4xxx/Kconfig.series:13]: Question @parthitce does this CPU have an FPU (floating point unit)? If so , please add this as well.

File: dts/arm/infineon/xmc4xxx.dtsi [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+#include <arm/armv7-m.dtsi>
+
+/ {
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu0: cpu@0 {
+			compatible = "arm,cortex-m4f";
```
Comment (ioannisg) [inline dts/arm/infineon/xmc4xxx.dtsi:16]: this entry suggests that this is M4F, so there's an FPU present

File: soc/arm/infineon_xmc/4xxx/Kconfig.series (hunk 9 above)
Comment (parthitce) [reply soc/arm/infineon_xmc/4xxx/Kconfig.series:13]: CPU_HAS_FPU was defined at Kconfig.soc and I just realized that all the XMC4xxx series is based on M4F. Moved to Kconfig.series. Thanks.

File: dts/arm/infineon/xmc4xxx.dtsi (hunk 10 above)
Comment (parthitce) [reply dts/arm/infineon/xmc4xxx.dtsi:16]: Yes, true.

**Comment by ioannisg** [comment]
@parthitce i guess you can fix also the 3rd and 4th checkpatch warning about lines exceeding 89 chars

**Review by ioannisg — APPROVED** [review]
Looks ok to me now. I guess if CI builds we can merge and any improvements/fixes needed can come during the freeze window.

**Comment by parthitce** [comment]
> @parthitce i guess you can fix also the 3rd and 4th checkpatch warning about lines exceeding 89 chars

@ioannisg Fixed, thanks.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@galak use of DT macros seems correct to me. I will merge this unless there's any objections once CI is green.

**Comment by galak** [comment]
> @galak use of DT macros seems correct to me. I will merge this unless there's any objections once CI is green.

Yeah, looks good to me as well.
